tests/test-upgrade.py

```diff
@@ -0,0 +1,76 @@
+import os
+import time
+from validators import validate_dns, validate_dashboard, validate_storage, validate_ingress
+from subprocess import check_call
+from utils import microk8s_enable, wait_for_pod_state, microk8s_disable, wait_for_installation
+
+upgrade_from = os.environ.get('UPGRADE_MICROK8S_FROM', 'beta')
+# Have UPGRADE_MICROK8S_TO point to a file to upgrade to that file
+upgrade_to = os.environ.get('UPGRADE_MICROK8S_TO', 'edge')
+
+
+class TestUpgrade(object):
+    """
+    Validates a microk8s upgrade path
+    """
+
+    def test_upgrade(self):
+        """
+        Deploy, probe, upgrade, validate nothing broke.
+
+        """
+        print("Testing upgrade from {} to {}".format(upgrade_from, upgrade_to))
+
+        cmd = "sudo snap install microk8s --classic --{}".format(upgrade_from).split()
+        check_call(cmd)
+        wait_for_installation()
+
+        # Run through the validators and
+        # select those that were valid for the original snap
+        test_matrix = {}
+        try:
+            microk8s_enable("dns")
+            wait_for_pod_state("", "kube-system", "running", label="k8s-app=kube-dns")
+            validate_dns()
+            test_matrix['dns'] = validate_dns
```

Comment:
It seems like these are running the validate functions twice, once here before adding them to the matrix and then once below when looping through the matrix.  I think the first line (`validate_dns()`) should be removed.

Reply:
The first run of the validation functions is to select those valid for the original snap. If they are not valid for the original snap there is no way to break them during an upgrade so there is no point in testing them after the upgrade.

Let me give you an example. The storage addon is available in edge but not in beta. When we decide to release a revision to beta all beta deployments will upgrade_from beta to the revision that is currently in edge. What we want to do is to snap install the beta snap, run through the validations and see which ones apply, upgrade_to edge and see if those valid validators still succeed. In the case of 'storage' the validation will fail in the beta snap so the upgrade will not break it either way and after upgrading to edge there is no point in testing the storage validation.